File: frontends/ts/src/adapter/VoiceEvent.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 RunAnywhere AI, Inc.

export type VoiceEvent =
  | { kind: 'user-said';        text: string; isFinal: boolean }
  | { kind: 'assistant-token';  text: string; tokenKind: TokenKind; isFinal: boolean }
  | { kind: 'audio';            pcm: Uint8Array; sampleRateHz: number }
  | { kind: 'interrupted';      reason: string }
  | { kind: 'state-change';     previous: PipelineState; current: PipelineState }
  | { kind: 'metrics';          latencyMs: number }
  | { kind: 'error';            code: number; message: string };

export type TokenKind = 'answer' | 'thought' | 'tool-call';

export type PipelineState =
  | 'idle' | 'listening' | 'thinking' | 'speaking' | 'stopped';

export class RunAnywhereError extends Error {
  constructor(public readonly code: number, message: string) {
    super(`[${code}] ${message}`);
    this.name = 'RunAnywhereError';
  }
}

File: sdk/dart/lib/adapter/voice_event.dart
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 RunAnywhere AI, Inc.

import 'dart:typed_data';

sealed class VoiceEvent {
  const VoiceEvent();
}

class UserSaid extends VoiceEvent {
  final String text;
  final bool   isFinal;
  const UserSaid(this.text, {required this.isFinal});
}

class AssistantToken extends VoiceEvent {
  final String    text;
  final TokenKind kind;
  final bool      isFinal;
  const AssistantToken(
    this.text, {
    this.kind = TokenKind.answer,
    required this.isFinal,
  });
}

class AudioFrame extends VoiceEvent {
  final Uint8List pcm;
  final int       sampleRateHz;
  const AudioFrame(this.pcm, this.sampleRateHz);
}

class Interrupted extends VoiceEvent {
  final String reason;
  const Interrupted(this.reason);
}

class VoiceError extends VoiceEvent {
  final int    code;
  final String message;
  const VoiceError(this.code, this.message);
}

enum TokenKind { answer, thought, toolCall }
